Implementation of list_contains

Dtenwolde opened the pull request:
As mentioned [here](https://github.com/duckdb/duckdb/issues/1940#issuecomment-871655466), one of the features yet to be implemented was the list_contains (or array_contains) function for arrays. This PR implements it for integers, floats and strings. Nested types are not supported at this moment. Behaviour is similar to PrestoDB. Happy to hear feedback @lnkuiper :slightly_smiling_face:

Commit:
- Started on list contains function
- Added first simple test of list contains

Commit:
- Tried to retrieve value, unsuccesful so far

Commit:
- Simplest test case now works for list_contains

Commit:
- Added a number of test cases
- Now works for integers and strings

Commit:
- Added more test cases, specifically NULL, double, string

Commit:
- Added more test for NULL and mixed number types (decimal / integer)

Commit:
- Added more test for NULL and mixed number types (decimal / integer)

Commit:
Removed attempt at supporting nested types, instead throw exception
Inverted some if-statement to improve readability.

Commit:
Now works with multiple lists for integers / strings

Commit:
Fixed adding list_contains.cpp to the wrong CMakeLists.txt file
Added const to variables where applicable
Formatted file

lnkuiper commented on src/function/scalar/list/list_contains.cpp:
```diff
@@ -0,0 +1,226 @@
+	return;
+}
+
+template <class T>
```
This function does not have to be a templated function if I understand correctly. This function only deals with `string_t`.

lnkuiper commented on src/function/scalar/list/list_contains.cpp:
```diff
@@ -0,0 +1,226 @@
+	auto result_entries = FlatVector::GetData<bool>(result); // Create a vector of bool
+	auto &result_validity = FlatVector::Validity(result);
+
+	for (idx_t i = 0; i < count; i++) {
```
This loop here is the "meat" of the code, which requires templating. The code that comes before is duplicated for the templated and the string code. I think there's a way to not have to duplicate this code

lnkuiper commented on src/function/scalar/list/list_contains.cpp:
```diff
@@ -0,0 +1,226 @@
+				continue;
+			}
+			auto actual_value = child_value[value_idx];
+			if (StringComparisonOperators::EqualsOrNot<false>(actual_value, ((T *)value_data.data)[0])) {
```
This line is the only difference between the templated and the string_t version of the code.

I think you can have only 1 `TemplatedListContainsFunction`, and use a templated comparison here. Then, all you need to do is write an explicit specialization for strings, e.g.:
```c++
template <class T>
static inline bool MyComparisonFunction(const T &left, const T&right) {
    return left == right;
}

template <>
inline bool MyComparisonFunction(const string_t &left, const string &right) {
    return StringComparisonOperators::EqualsOrNot<false>(left, right);
}
```

This saves a full duplication of the code.

lnkuiper commented on src/function/scalar/list/list_contains.cpp:
```diff
@@ -0,0 +1,226 @@
+				continue;
+			}
+			auto actual_value = child_value[value_idx];
+			if (actual_value == ((T *)value_data.data)[0]) {
```
Instead of doing:
```c++
(T *)value_data.data)[0]
```
Every iteration, you can do this before the loop:
```c++
auto values = (T *)value_data.data;
```

And then index with
```c++
values[0]
```

lnkuiper commented on test/sql/function/list/list_contains.test:
```diff
@@ -0,0 +1,190 @@
+create table TEST (i int[]);
+
+statement ok
+insert into TEST values ([2,1,3]), ([2,3,4]), ([1]);
```
Can we also get a test case with a non-constant value, e.g.:
```sql
statement ok
create table TEST2 (i int[], int);

statement ok
insert into TEST2 values ([2,1,3], 2), ([2,3,4], 5), ([1], NULL);

query T
select list_contains(i, j) from TEST2
----
1
0
NULL
```

I suspect this does not work yet

lnkuiper commented on src/function/scalar/list/list_contains.cpp:
```diff
@@ -0,0 +1,226 @@
+	case PhysicalType::LIST:
+		throw NotImplementedException("This function has not yet been implemented for nested types");
+	default:
+		throw InvalidTypeException(args.data[1].GetType().id(), "Invalid type for List Contains");
```
We can check whether nested types have been supplied in `ListContainsBind` already, that way the errors happens before the query has started executing. This should also be a `NotImplementedException`, because I don't think any type should be invalid for this function

lnkuiper reviewed (changes requested):
Great job that you got this to work! I know from experience that it's challenging to start working on these things, especially if they involve nested types :)

Some comments:

Commit:
Now uses templated comparison, removing the TemplatedListContainsStringFunction

Commit:
Now works with two columns as argument

Commit:
Removed SetResultFalse as it wasn't needed anymore.
Removed edge case if-statements as they are handled correctly in another way now.
Now also works with nested types: List, Map, and Struct. Added test cases for those

Dtenwolde commented:
Implemented the feedback provided. 
Now also works with nested types (list, map and struct)

Commit:
Improving code coverage for float and double

Dtenwolde commented:
The code coverage mentions three places that are uncovered, however using the debugger I see that those points are reached. I'm not sure which tests to add to make sure codecov also thinks these points are covered.

lnkuiper commented on src/function/scalar/list/list_contains.cpp:
```diff
@@ -0,0 +1,225 @@
+			if (!child_data.validity.RowIsValid(child_value_idx)) {
+				continue;
+			}
+			if (ValueCompare(child_value[child_value_idx], values[value_index])) {
```
There are some issues with the Windows CI, and I suspect this has something to do with not specifying the template class.

Can you try to change this to `ValueCompare<T>` ?

lnkuiper commented on src/function/scalar/list/list_contains.cpp:
```diff
@@ -0,0 +1,225 @@
+			if (!child_data.validity.RowIsValid(child_value_idx)) {
+				continue;
+			}
+			if (ValueCompare(child_vector.GetValue(child_value_idx), value_vector.GetValue(value_index))) {
```
Same here, try `ValueCompare<Value>`

lnkuiper commented on src/function/scalar/list/list_contains.cpp:
```diff
@@ -0,0 +1,225 @@
+}
+
+template <class T>
+static void NestedTemplatedListContainsFunction(DataChunk &args, ExpressionState &state, Vector &result) {
```
This method does not need to be templated, because we are using Value. Windows CI issues may also relate to this.

lnkuiper commented on test/sql/function/list/list_contains.test:
```diff
@@ -0,0 +1,358 @@
+create table STR_TEST (i string[]);
+
+statement ok
+insert into STR_TEST values (['a','b','c']), (['d','a','e']), (['b']);
```
Just to be sure, but can you also create a test that has strings longer than 12 characters? DuckDB handles these differently.

lnkuiper reviewed (changes requested):
Great work! I'm happy the nested types work too.

Just a few minor things to solve the CI issues, then it is ready to merge.

Commit:
Added test with longer strings
Added <Value> and <T> with ValueCompare call
Removed template from NestedListContainsFunction

lnkuiper commented on src/function/scalar/list/list_contains.cpp:
```diff
@@ -0,0 +1,224 @@
+#include <iostream>
```
Can this import be removed?

lnkuiper commented on src/function/scalar/list/list_contains.cpp:
```diff
@@ -0,0 +1,224 @@
+			}
+		}
+	}
+	return;
```
the return here and in the nested function are not needed

lnkuiper commented on src/function/scalar/list/list_contains.cpp:
```diff
@@ -0,0 +1,224 @@
+		auto list_index = list_data.sel->get_index(i);
+		auto value_index = value_data.sel->get_index(i);
+
+		if (!list_data.validity.RowIsValid(list_index) or !value_data.validity.RowIsValid(value_index)) {
```
I think this `or` here is the culprit of the Windows CI failing. Can you change this to `||`?

lnkuiper commented on src/function/scalar/list/list_contains.cpp:
```diff
@@ -0,0 +1,224 @@
+		auto list_index = list_data.sel->get_index(i);
+		auto value_index = value_data.sel->get_index(i);
+
+		if (!list_data.validity.RowIsValid(list_index) or !value_data.validity.RowIsValid(value_index)) {
```
Same here, `or` to `||`

lnkuiper reviewed (changes requested):
Some minor nitpicks.

The Windows CI is still failing, so we cannot merge this yet. Last time I suggested changing some template specifications, but that did not fix it.
The output is cryptic as always: https://github.com/duckdb/duckdb/runs/5218814475?check_suite_focus=true#step:4:546

I think I found it, though. See my comment about `or`!

Commit:
Changed or to ||
removed unused imports and useless returns

lnkuiper commented:
Looks Windows is happy now. The CI failures seem unrelated.

shanzi commented on src/function/scalar/list/list_contains.cpp:
```diff
@@ -0,0 +1,221 @@
+	} else if (list.id() == LogicalTypeId::SQLNULL || value.id() == LogicalTypeId::SQLNULL) {
+		// In case either the list or the value is NULL, return NULL
+		// Similar to behaviour of prestoDB
+		bound_function.arguments[0] = list;
```
I'm wondering  should we follow presto's behavior here as NULL seems to be a valid value in a list. ex.
```sql
select list_contains([1, 2, NULL], 1)
```
Will output `true`.
Thus, a user may expect
```sql
select list_contains([1, 2, NULL], NULL)
```
Also to be true.
